File: orchestrator/dashboard/pricing.py
"""Pricing table loader + fallback cost calculator.

Contract (from the MVP brief):
    - The dashboard trusts `SpendEntry.cost_usd` when it's a real positive
      number. That's what the orchestrator recorded at dispatch time and
      it's the source of truth (claude/anthropic return real billing).
    - When `cost_usd` is 0 (or None / missing / malformed), we fall back
      to this table to *estimate* a cost from tokens. This matters for
      opencode and codex where there's no billing API and the recorded
      cost is always 0.
    - The table lives at `dashboard/pricing.yaml` (defaults) and can be
      overridden per project with a `pricing.yaml` at `<project_root>/`.

The loader is intentionally forgiving: missing YAML, missing keys, missing
`default` — none of them crash. Worst case we return 0.0 and the dashboard
shows "$0.00" for that row.

Prices are USD per 1M tokens. Formula:
    cost = tokens_in * input / 1_000_000 + tokens_out * output / 1_000_000
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any


_DEFAULT_TABLE_PATH = Path(__file__).parent / "pricing.yaml"

# Absolute floor when nothing else resolves. We prefer to under-estimate than
# to invent a $99/1M price out of thin air.
_HARDCODED_FALLBACK: dict[str, float] = {"input": 1.00, "output": 4.00}


@dataclass(frozen=True, slots=True)
class ModelPrice:
    """Per-1M-tokens price for a single model."""

    input: float
    output: float

    def cost(self, tokens_in: int, tokens_out: int) -> float:
        """Compute USD cost for a dispatch. Negative tokens → 0."""
        ti = max(0, int(tokens_in or 0))
        to = max(0, int(tokens_out or 0))
        return (ti * self.input + to * self.output) / 1_000_000.0


class PricingTable:
    """Layered price table: project overrides on top of dashboard defaults.

    Layering rules:
        - Both files are optional. Missing files are silently ignored.
        - Project keys REPLACE (not merge with) default keys of the same
          model name — you can't tweak just `input` and keep `output`. This
          keeps the mental model simple: one row = one full price.
        - Model lookup order: exact key → `default` key → hardcoded floor.
    """

    def __init__(self, prices: dict[str, ModelPrice]):
        self._prices = prices

    # ---- constructors ---------------------------------------------------

    @classmethod
    def load(cls, project_root: Path | None = None) -> "PricingTable":
        """Load defaults from the packaged YAML, then layer project overrides.

        `project_root` is optional — when None (or when no `pricing.yaml`
        exists there), only defaults are used.
        """
        merged: dict[str, dict[str, float]] = {}
        merged.update(_read_models(_DEFAULT_TABLE_PATH))
        if project_root is not None:
            override = project_root / "pricing.yaml"
            merged.update(_read_models(override))

        # Guarantee a `default` row so `for_model` always finds SOMETHING
        # short of the hardcoded floor. Cheap safety net for typos in YAML.
        merged.setdefault("default", dict(_HARDCODED_FALLBACK))

        prices = {
            name: ModelPrice(
                input=float(row.get("input", 0.0) or 0.0),
                output=float(row.get("output", 0.0) or 0.0),
            )
            for name, row in merged.items()
        }
        return cls(prices)

    # ---- lookup ---------------------------------------------------------

    def for_model(self, model: str) -> ModelPrice:
        """Return the price for `model`, falling back to `default` → floor."""
        if model in self._prices:
            return self._prices[model]
        if "default" in self._prices:
            return self._prices["default"]
        return ModelPrice(**_HARDCODED_FALLBACK)

    def has(self, model: str) -> bool:
        """True if `model` has an explicit row (not just the default fallback)."""
        return model in self._prices

    def estimate_cost(self, model: str, tokens_in: int, tokens_out: int) -> float:
        """Convenience: look up + compute in one call."""
        return self.for_model(model).cost(tokens_in, tokens_out)

    def resolve_cost(
        self,
        recorded_cost: float | None,
        model: str,
        tokens_in: int,
        tokens_out: int,
    ) -> float:
        """Return the cost we should display for a spend row.

        Rules:
            - `recorded_cost` > 0                → use it (source of truth)
            - `recorded_cost` in {0, None, NaN}  → estimate from pricing table
            - malformed model                    → fall back to `default` row
        """
        try:
            rc = float(recorded_cost) if recorded_cost is not None else 0.0
        except (TypeError, ValueError):
            rc = 0.0
        if rc > 0:
            return rc
        return self.estimate_cost(model or "default", tokens_in, tokens_out)

    # ---- introspection --------------------------------------------------

    def models(self) -> list[str]:
        """All model keys currently in the table, sorted for stable UIs."""
        return sorted(self._prices.keys())


# ---- helpers ----------------------------------------------------------------


def _read_models(path: Path) -> dict[str, dict[str, float]]:
    """Read a `pricing.yaml` file and return its `models:` map.

    Missing file → `{}`. Malformed YAML → `{}` with no crash. `models:` is
    the only accepted top-level shape; anything else is treated as empty.
    """
    if not path.exists():
        return {}
    try:
        import yaml
    except ImportError:  # pragma: no cover — pyyaml is a hard dep of orch
        return {}
    try:
        with path.open("r", encoding="utf-8") as fh:
            data: Any = yaml.safe_load(fh) or {}
    except Exception:  # noqa: BLE001 — corrupt YAML shouldn't kill the dashboard
        return {}
    if not isinstance(data, dict):
        return {}
    models = data.get("models") or {}
    if not isinstance(models, dict):
        return {}
    # Filter: keep only rows that look like `{input: float, output: float}`.
    out: dict[str, dict[str, float]] = {}
    for name, row in models.items():
        if not isinstance(row, dict):
            continue
        out[str(name)] = {
            "input": float(row.get("input", 0.0) or 0.0),
            "output": float(row.get("output", 0.0) or 0.0),
        }
    return out

File: internal/pricing/pricing.go
// Package pricing is the fallback price table the dashboard uses to put a
// number on a dispatch nobody billed for.
//
// Ported from `orchestrator/dashboard/pricing.py`. The contract is the part
// worth keeping in mind: a recorded `cost_usd` above zero is the truth and is
// never second-guessed — that is what the provider charged. The table is only
// consulted when the recorded cost is zero, which is the normal case for
// opencode and local codex, where there is no billing API to ask.
package pricing

import (
	"embed"
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"gopkg.in/yaml.v3"
)

//go:embed defaults/pricing.yaml
var defaultsFS embed.FS

// hardcodedFallback is the floor when nothing else resolves — deliberately
// low. Under-estimating a cost is a smaller lie than inventing a $99/1M price
// for a model the table has never heard of.
var hardcodedFallback = ModelPrice{Input: 1.00, Output: 4.00}

// ModelPrice is USD per million tokens.
type ModelPrice struct {
	Input  float64
	Output float64
}

// Cost is what a dispatch of this size costs. Negative token counts are read
// as zero rather than as a refund.
func (p ModelPrice) Cost(tokensIn, tokensOut int) float64 {
	if tokensIn < 0 {
		tokensIn = 0
	}
	if tokensOut < 0 {
		tokensOut = 0
	}
	return (float64(tokensIn)*p.Input + float64(tokensOut)*p.Output) / 1_000_000.0
}

// Table is the layered price list: a project's own `pricing.yaml` on top of
// the packaged defaults.
//
// A project key REPLACES the default row of the same name rather than merging
// into it — you cannot tweak `input` and inherit `output`. One row is one
// complete price, which is the only version of this anybody can hold in their
// head while reading a YAML file.
type Table struct {
	prices map[string]ModelPrice
}

// Load reads the embedded defaults and layers a project override on top.
//
// Every failure is silent and local: a missing file, unreadable YAML, a row
// that is not a price. A dashboard that refused to render because somebody
// hand-edited `pricing.yaml` would be trading a wrong number for no page at
// all, and this table's whole purpose is being an estimate.
//
// projectRoot may be empty, meaning "defaults only".
func Load(projectRoot string) Table {
	merged := map[string]ModelPrice{}

	body, err := defaultsFS.ReadFile("defaults/pricing.yaml")
	if err == nil {
		for name, price := range parseModels(body) {
			merged[name] = price
		}
	}
	if projectRoot != "" {
		// #nosec G304 -- the project's own pricing.yaml, named by the operator
		if override, readErr := os.ReadFile(filepath.Join(projectRoot, "pricing.yaml")); readErr == nil {
			for name, price := range parseModels(override) {
				merged[name] = price
			}
		}
	}
	// Guarantee a `default` row so ForModel always finds something short of
	// the floor. Cheap insurance against a typo in the shipped YAML.
	if _, ok := merged["default"]; !ok {
		merged["default"] = hardcodedFallback
	}
	return Table{prices: merged}
}

// ForModel resolves a price: the exact name, then `default`, then the floor.
func (t Table) ForModel(model string) ModelPrice {
	if p, ok := t.prices[model]; ok {
		return p
	}
	if p, ok := t.prices["default"]; ok {
		return p
	}
	return hardcodedFallback
}

// Has reports whether the model has a row of its own, as opposed to falling
// through to `default`.
func (t Table) Has(model string) bool {
	_, ok := t.prices[model]
	return ok
}

// EstimateCost looks up and computes in one call.
func (t Table) EstimateCost(model string, tokensIn, tokensOut int) float64 {
	return t.ForModel(model).Cost(tokensIn, tokensOut)
}

// ResolveCost is the rule the whole package exists for: a recorded cost above
// zero wins, and anything else — zero, missing, negative — is estimated from
// tokens.
//
// Note that a recorded cost of exactly zero is treated as "not recorded", not
// as "free". That is Python's rule and it is the right one here: the backends
// that genuinely cost nothing to call are the same ones that report no cost at
// all, so the two are indistinguishable in the row.
func (t Table) ResolveCost(recordedCost float64, model string, tokensIn, tokensOut int) float64 {
	if recordedCost > 0 {
		return recordedCost
	}
	if model == "" {
		model = "default"
	}
	return t.EstimateCost(model, tokensIn, tokensOut)
}

// Models lists every key, sorted, for a UI that wants to show the table.
func (t Table) Models() []string {
	out := make([]string, 0, len(t.prices))
	for name := range t.prices {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

// parseModels reads the `models:` map out of one pricing.yaml.
//
// Anything that is not that shape is nothing: a document that is not a
// mapping, a `models` that is not a mapping, a row that is not a mapping. The
// keys that survive are forced to floats, so `input: "3.00"` — which a YAML
// file written by hand will have — still reads as 3.
func parseModels(body []byte) map[string]ModelPrice {
	var doc struct {
		Models map[string]map[string]any `yaml:"models"`
	}
	if err := yaml.Unmarshal(body, &doc); err != nil {
		return nil
	}
	out := make(map[string]ModelPrice, len(doc.Models))
	for name, row := range doc.Models {
		if row == nil {
			continue
		}
		out[name] = ModelPrice{
			Input:  asFloat(row["input"]),
			Output: asFloat(row["output"]),
		}
	}
	return out
}

func asFloat(raw any) float64 {
	switch v := raw.(type) {
	case float64:
		return v
	case int:
		return float64(v)
	case string:
		var f float64
		if _, err := fmt.Sscanf(v, "%g", &f); err == nil {
			return f
		}
	}
	return 0
}
